### PR TITLE
U4-S2 Cai Nowicki

### PR DESCRIPTION
**Pull request by CaiNowicki**

### CODE REVIEW

**Comment by llpk79** [comment]
# Data Science Unit 4 Sprint 2 Assignment 1

Define the Following:
- Nice work.

Write your own perceptron code that can correctly classify (99.0% accuracy) a NAND gate.
- Was finally able to get the first bit to work by adding more rows and a 'noise' column.
  ```inputs = np.array([
    [0, 0, 1],
    [1, 0, 0],
    [0, 1, 1],
    [1, 1, 0],
    [1, 0, 1],
    [0, 1, 0],
    [1, 1, 0],
    [1, 0, 1],])  
    correct_outputs = np.array([[1], [1], [1], [0], [1], [1], [0], [1]])

Implement your own Perceptron Class and use it to classify a binary dataset:
- Research how to add a bias term to your perceptron and see if that will improve results.

**Comment by llpk79** [comment]
# Data Science Unit 4 Sprint 2 Assignment 2

### Implement a 3 input, 4 node hidden-layer, 1 output node Multilayer Perceptron on the following dataset:

In `class NeuralNetwork` in `def backward`:
- In my version I take the sigmoid derivative of the activated outputs. 
```   
  def backpass(self, X, y, out):
        y = y.reshape((y.shape[0], 1))
        error = y - out
        
        self.errors.append(np.sum(error**2))
        # Caluculate adjustment from hidden -> output.
        delta_output = error * self.sigmoid_prime(out) <--
        
        # Calculate error from input -> hidden.
        output_error = delta_output.dot(self.output_weight[1:].T)
        delta_hidden = output_error * self.sigmoid_prime(self.activated_hidden) <--
```
- Both appear to work, but mine gets to `if loss < 0.0001:` in about half as many epochs.

### Try building/training a more complex MLP on a bigger dataset.
- Need to rework the architecture here to accept a flattened version of images, then output a softmax activation with ten output nodes for your ten classes. A challenge I'm sure you're up to! 😄

**Comment by llpk79** [comment]
# Data Science Unit 4 Sprint 2 Assignment 3

### Use the Keras Library to build a Multi-Layer Perceptron Model on the Boston Housing dataset
- Don't forget to normalize the data.

### Use the Keras Library to build an image recognition network using the Fashion-MNIST dataset (also comes with keras)
- Was able to get `val_accuracy` to `0.8858` by normalizing(`X_train = X_train / 255`) and onehot encoding (change `loss` to `loss='categorical_crossentropy'`) in under 10 epochs without changing your architecture.

#### Nice work getting your image into the mix!!!
